floodmodeller_api/toolbox/model_build/structure_log/structure_log.py

```diff
@@ -19,237 +51,446 @@ def _add_fields(self):
+                "orifice_flow",
+                "orifice_lower_transition_dist",
+                "orifice_upper_transition_dist",
+                "orifice_discharge_coefficient",
+            },
+        )
+        orifice_data["transition_width"] = (
+            structure.orifice_upper_transition_dist + structure.orifice_lower_transition_dist
+        )
+
+        opening_data = []
+        if (
+            structure.opening_nrows > 0
+        ):  # if it has openings, not sure why it wouldnt but worth checking.
+            opening_df = structure.opening_data
+            for (
+                _,
+                row,
+            ) in (
+                opening_df.iterrows()
+            ):  # this isnt 'proper' for a df but there should be few rows and we're doing special stuff
+                opening = {}
+                opening["width"] = (
+                    row["Finish"] - row["Start"]
+                )  # round here because fp precision looks uncool
+
+                # this is crude and can be wrong where the min is on the 'edge' of the section instead of at one of the points, but should be correct 99.9% of the time.
+                temp_df = section_df.loc[
+                    (row["Start"] <= section_df["X"]) & (section_df["X"] <= row["Finish"])
+                ]
+                opening["bed_minimum"] = temp_df["Y"].min()
+
+                opening["soffit_level"] = row["Soffit Level"]
+                opening["opening_height"] = opening["soffit_level"] - opening["bed_minimum"]
+
+                opening_data.append(opening)
+
+        culvert_data = []
+        if hasattr(structure, "culvert_data") and structure.culvert_data.shape[0] > 1:
+            for _, row in structure.culvert_data:
+                culvert = {
+                    "invert": row["Invert"],
+                    "soffit": row["Soffit"],
+                    "bore_area": "Section Area",
+                }
+                culvert["height"] = culvert["soffit"] - culvert["invert"]
+                culvert["average_width"] = culvert["bore_area"] / culvert["height"]
+                culvert_data.append(culvert)
+
+        return {
+            "opening_data": opening_data,
+            "friction": friction,
+            "orifice_data": orifice_data,
+            "culvert_data": culvert_data,
+        }
+
+    def _sluice_data(self, structure):  # these could do with more attention, given more time
+        dimensions = extract_attrs(structure, {"crest_elevation", "weir_breadth", "weir_length"})
+
+        return {"dimensions": dimensions}
+
+    def _weir_data(self, structure):
+        dimensions = extract_attrs(structure, {"weir_elevation", "weir_breadth"})
+
+        if structure._unit == "RNWEIR":
+            dimensions |= extract_attrs(
+                structure,
+                {"weir_length", "upstream_crest_height", "downstream_crest_height"},
+            )
+
+        return {"dimensions": dimensions}
 
     def _add_structures(self):
```

Comment:
Cant remember if I commented this elsewhere, this could probably do with a refactor and would also help with https://github.com/People-Places-Solutions/floodmodeller-api/pull/141/files#r1831033444 , although it might be better if we leave as-is for now, and register it as an issue to be handled later